README.md
# Hidraw

Hidraw is an Elixir interface to Linux hidraw devices.

## Usage

Hidraw can be used to monitor `/dev/hidraw*` devices and report raw data to
the parent process.

Start by looking for the device you want to monitor.

```elixir
iex> Hidraw.enumerate
[
  {"/dev/hidraw2", "ﾩSymbol Technologies, Inc, 2002 Symbol Bar Code Scanner"},
  {"/dev/hidraw1", "DLL082A:01 06CB:76AF"},
  {"/dev/hidraw0", "ELAN Touchscreen"}
]
```

Here we have a Barcode scanner at `/dev/hidraw2`, lets open it.

```elixir
iex> Hidraw.start_link "/dev/hidraw2"
{:ok, #PID<0.197.0>}
```

The first message we will receive is the device's report descriptor

```elixir
iex> flush
{:hidraw, "/dev/hidraw2",
 {:report_descriptor,
  <<6, 69, 255, 10, 0, 75, 161, 1, 10, 1, 74, 117, 8, 149, 11, 21, 0, 38, 255,
    0, 145, 2, 10, 2, 74, 149, 64, 129, 2, 192>>}}
```

All subsequent messages will be triggered off device events.
Here I am scanning a barcode:

```elixir
iex(4)> flush
{:hidraw, "/dev/hidraw2",
 <<16, 16, 3, 0, 65, 67, 67, 49, 55, 49, 49, 50, 79, 0, 24, 11, 0, 0, 0, 0, 0,
   0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
   ...>>}
```


CHANGELOG.md
# Hidraw

## v0.2.2

* Bug Fixes
  * Move C build files to `_build`

## v0.2.1

* Bug Fixes
  * Update files in hex package

## v0.2.0

* Enhancements
  * Format code.
  * Publish to hex

## v0.1.1

* Enhancements
  * Fixed warnings
  * Compile using suid bit for discovering devices on linux machines


test/hidraw_test.exs
defmodule HidrawTest do
  use ExUnit.Case
  doctest Hidraw

  test "the truth" do
    assert 1 + 1 == 2
  end
end


mix.exs
defmodule Hidraw.MixProject do
  use Mix.Project

  @app :hidraw

  def project do
    [
      app: @app,
      version: "0.2.2",
      elixir: "~> 1.4",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      make_targets: ["all"],
      make_clean: ["clean"],
      compilers: [:elixir_make | Mix.compilers()],
      deps: deps(),
      docs: [extras: ["README.md"], main: "readme"]
    ]
  end

  def application do
    [applications: [:logger]]
  end

  defp deps do
    [
      {:elixir_make, "~> 0.6"},
      {:ex_doc, "~> 0.18", only: :dev}
    ]
  end

  defp description do
    "Elixir interface to Linux hidraw devices"
  end

  defp package do
    [
      files: ["lib", "LICENSE", "mix.exs", "README.md", "src/*.[ch]", "Makefile"],
      licenses: ["Apache 2.0"],
      links: %{"Github" => "https://github.com/letoteteam/#{@app}"}
    ]
  end
end


lib/hidraw.ex
defmodule Hidraw do
  use GenServer

  def start_link(fd) do
    GenServer.start_link(__MODULE__, [fd, self()])
  end

  def enumerate() do
    executable = :code.priv_dir(:hidraw) ++ '/hidraw'

    port =
      Port.open({:spawn_executable, executable}, [
        {:args, ["enumerate"]},
        {:packet, 2},
        :use_stdio,
        :binary
      ])

    receive do
      {^port, {:data, <<?r, message::binary>>}} ->
        :erlang.binary_to_term(message)
    after
      5_000 ->
        Port.close(port)
        []
    end
  end

  def report_descriptor(pid, timeout \\ 5000) do
    GenServer.call(pid, :report_desc, timeout)
  end

  def init([fd, caller]) do
    executable = :code.priv_dir(:hidraw) ++ '/hidraw'

    port =
      Port.open({:spawn_executable, executable}, [
        {:args, [fd]},
        {:packet, 2},
        :use_stdio,
        :binary,
        :exit_status
      ])

    state = %{port: port, name: fd, callback: caller, buffer: [], report_desc: nil}

    {:ok, state}
  end

  def handle_call(:report_desc, _from, s) do
    {:reply, {:ok, s.report_desc}, s}
  end

  def handle_info({_, {:data, <<?n, message::binary>>}}, state) do
    msg = :erlang.binary_to_term(message)
    handle_port(msg, state)
  end

  def handle_info({_, {:data, <<?d, message::binary>>}}, state) do
    {:descriptor, descriptor} = :erlang.binary_to_term(message)
    send(state.callback, {:hidraw, state.name, {:report_descriptor, descriptor}})
    {:noreply, %{state | report_desc: descriptor}}
  end

  def handle_info({_, {:data, <<?e, message::binary>>}}, state) do
    error = :erlang.binary_to_term(message)
    send(state.callback, {:hidraw, state.name, error})
    {:stop, error, state}
  end

  def handle_info({_, {:exit_status, status}}, state) do
    send(state.callback, {:hidraw, state.name, {:exit, status}})
    {:stop, {:exit, status}, state}
  end

  defp handle_port({:data, value}, state) do
    send(state.callback, {:hidraw, state.name, value})
    {:noreply, state}
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :hidraw, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:hidraw, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


src/hidraw_enum.c
/*
 *  Copyright 2016 Justin Schneck
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "hidraw_enum.h"

#include <linux/types.h>
#include <linux/input.h>
#include <linux/hidraw.h>

#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <dirent.h>
#include <inttypes.h>

#include "utils.h"
#include "erlcmd.h"

static const char response_id = 'r';

struct device_info *device_info_alloc()
{
    struct device_info *info = (struct device_info *) malloc(sizeof(struct device_info));
    memset(info, 0, sizeof(struct device_info));
    return info;
}

void device_info_free(struct device_info *info)
{
    // Free any data
    if (info->fd)
        free(info->fd);
    if (info->name)
        free(info->name);

    // Reset the fields
    memset(info, 0, sizeof(struct device_info));
}

void device_info_free_list(struct device_info *info)
{
    while (info) {
        struct device_info *next = info->next;
        device_info_free(info);
        free(info);
        info = next;
    }
}

/**
 * Filter for the AutoDevProbe scandir on /dev/input.
 *
 * @param dir The current directory entry provided by scandir.
 *
 * @return Non-zero if the given directory entry starts with "event", or zero
 * otherwise.
 */
static int is_event_device(const struct dirent *dir) {
	return strncmp(HIDRAW_DEV_NAME, dir->d_name, 5) == 0;
}

int enum_devices() {
  if (getuid() != 0)
	  fprintf(stderr, "Not running as root, no devices may be available.\n");

  struct device_info *device_list = find_hidraw_devices();
  int device_list_len = 0;
  for (struct device_info *device = device_list;
         device != NULL;
         device = device->next)
        device_list_len++;
  debug("Found %d devices", device_list_len);
  char resp[4096];
  int resp_index = sizeof(uint16_t); // Space for payload size
  resp[resp_index++] = response_id;
  ei_encode_version(resp, &resp_index);
  ei_encode_list_header(resp, &resp_index, device_list_len);

  for (struct device_info *device = device_list;
         device != NULL;
         device = device->next) {
    ei_encode_tuple_header(resp, &resp_index, 2);
    ei_encode_binary(resp, &resp_index, device->fd, strlen(device->fd));
    ei_encode_binary(resp, &resp_index, device->name,strlen(device->name));
  }
  ei_encode_empty_list(resp, &resp_index);
  erlcmd_send(resp, resp_index);
  device_info_free_list(device_list);
  return 0;
}

struct device_info *find_hidraw_devices() {
  struct dirent **namelist;
	int i, ndev, devnum;
	char *filename;

  struct device_info *info = NULL;

  ndev = scandir(DEV, &namelist, is_event_device, alphasort);
	if (ndev <= 0)
		return info;

  for (i = 0; i < ndev; i++)
	{

		char fname[64];
		int fd = -1;
		char name[256] = "???";

		snprintf(fname, sizeof(fname),
			 "%s/%s", DEV, namelist[i]->d_name);
		fd = open(fname, O_RDWR|O_NONBLOCK);
		if (fd < 0)
			continue;
    ioctl(fd, HIDIOCGRAWNAME(256), name);
    close(fd);

    struct device_info *new_info = device_info_alloc();

    new_info->fd = strdup(fname);
    new_info->name = strdup(name);

    new_info->next = info;
    info = new_info;

		free(namelist[i]);
	}
  free(namelist);
  return info;
}


src/hidraw_enum.h
/*
 *  Copyright 2016 Justin Schneck
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef HIDRAW_ENUM_H
#define HIDRAW_ENUM_H

#define DEV "/dev"
#define HIDRAW_DEV_NAME "hidraw"

struct device_info {
    char *fd;
    char *name;

    struct device_info *next;
};

struct device_info *device_info_alloc();
void device_info_free(struct device_info *info);
void device_info_free_list(struct device_info *info);

int enum_devices(void);
struct device_info *find_hidraw_devices();

#endif


src/ex_hidraw.c
/*
 *  Copyright 2016 Justin Schneck
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <poll.h>
#include <err.h>
#include <fcntl.h>
#include <linux/hidraw.h>
#include <linux/types.h>

#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "utils.h"
#include "erlcmd.h"
#include "hidraw_enum.h"

static const char notification_id = 'n';
static const char descriptor_id = 'd';
static const char error_id = 'e';

void device_handle_request(const char *req, void *cookie) {
  debug("Erl sent data");
}

void device_process(int fd) {
  debug("Read");
  int i, res = 0;
  char buf[1024];

  /* Get a report from the device */
  res = read(fd, buf, 1024);
  if (res < 0) {
    debug("read");
  } else {
    debug("read() read %d bytes:\n\t", res);
    char resp[res * 64];
    int resp_index = sizeof(uint16_t); // Space for payload size

    debug("%s", buf);

    resp[resp_index++] = notification_id;
    ei_encode_version(resp, &resp_index);
    ei_encode_tuple_header(resp, &resp_index, 2);
    ei_encode_atom(resp, &resp_index, "data");

    ei_encode_binary(resp, &resp_index, buf, res);
    erlcmd_send(resp, resp_index);
  }
}

void device_closed(int fd) {
  debug("device closed");
  char resp[16];
  int resp_index = sizeof(uint16_t); // Space for payload size

  resp[resp_index++] = error_id;

  ei_encode_version(resp, &resp_index);
  ei_encode_tuple_header(resp, &resp_index, 2);
  ei_encode_atom(resp, &resp_index, "error");
  ei_encode_atom(resp, &resp_index, "closed");
  erlcmd_send(resp, resp_index);
}

static int open_device(char *dev) {
  int fd;
  int i, res, desc_size = 0;
  char buf[256];

  struct hidraw_report_descriptor rpt_desc;
  struct hidraw_devinfo info;

  memset(&rpt_desc, 0x0, sizeof(rpt_desc));
  memset(&info, 0x0, sizeof(info));
  memset(buf, 0x0, sizeof(buf));

  debug("Open device %s", dev);
  fd = open(dev, O_RDWR|O_NONBLOCK);
  if (errno == EACCES && getuid() != 0)
    err(EXIT_FAILURE, "You do not have access to %s.", dev);

    /* Get Report Descriptor Size */
    res = ioctl(fd, HIDIOCGRDESCSIZE, &desc_size);
    if (res < 0)
      perror("HIDIOCGRDESCSIZE");
    else
      debug("Report Descriptor Size: %d\n", desc_size);
    /* Get Report Descriptor */
    rpt_desc.size = desc_size;
    res = ioctl(fd, HIDIOCGRDESC, &rpt_desc);
    if (res < 0) {
      perror("HIDIOCGRDESC");
    }

    debug("%s ", rpt_desc.value);

  struct erlcmd handler;
  erlcmd_init(&handler, device_handle_request, &fd);

  char resp[desc_size * 64];
  int resp_index = sizeof(uint16_t); // Space for payload size
  resp[resp_index++] = descriptor_id;
  ei_encode_version(resp, &resp_index);
  ei_encode_tuple_header(resp, &resp_index, 2);
  ei_encode_atom(resp, &resp_index, "descriptor");

  ei_encode_binary(resp, &resp_index, rpt_desc.value, desc_size);
  erlcmd_send(resp, resp_index);

  for (;;) {
    struct pollfd fdset[2];

    fdset[0].fd = STDIN_FILENO;
    fdset[0].events = POLLIN;
    fdset[0].revents = 0;

    fdset[1].fd = fd;
    fdset[1].events = (POLLIN | POLLPRI | POLLHUP);
    fdset[1].revents = 0;

    int timeout = -1; // Wait forever unless told by otherwise
    int rc = poll(fdset, 2, timeout);

    if (fdset[0].revents & (POLLIN | POLLHUP))
      erlcmd_process(&handler);

    if (fdset[1].revents & POLLIN)
      device_process(fd);

    if (fdset[1].revents & POLLHUP) {
      device_closed(fd);
      break;
    }
  }
  debug("Exit");
  return 0;
}

int main(int argc, char *argv[]) {
  if (argc == 2 && strcmp(argv[1], "enumerate") == 0)
    return enum_devices();
  else
    return open_device(strdup(argv[argc - 1]));
}
